### DISCUSSION TITLE
C-like API?

### DISCUSSION THREAD

**Discussion by siiky** [zxing-cpp/zxing-cpp#551] [Ideas]
**EDIT**: A C-API is now available as a "technology preview" under `wrappers/c` (see #553).

Hi all,

I recently had to use ZXing in Go (which only officially supports C external libraries through CGo) and the process wasn't the most straightforward possible. After a long time of confusion, ZXing is actually extremely easy and simple to use! But I had to cut a lot of corners I would prefer not to have cut... :/

So, because of that, I would like to suggest/ask for your opinion on an "official" C API.

Basically I envision a single `.h` wrapper file that exports C names for ZXing functions, plus an accompanying `.cpp` with the wrapper function bodies (or maybe a single-header-library à la stb &c would also work). Something like this:

```cpp
// zxing.h

#ifdef __cplusplus
extern "C" {
#endif

  zxing_ImageView * zxing_ImageView_new (void * data, int width, int height, zxing_ImageFormat fmt);
  zxing_Result * zxing_ReadBarcode (zxing_ImageView * iv, zxing_DecodeHints * hints);
  // ...

#ifdef __cplusplus
}
#endif
```

This would make the life of other people trying to use ZXing in C (or languages that only support C) much easier and more straightforward.

One difficulty is some C++ features that are "C-unfriendly", such as the `Type` enum of the `Error` type. I'm not very familiar with C++, so maybe it's possible, but I don't see how to define the equivalent enum in a plain C header file. This is the best I know.

```c
enum zxing_Error_Type {
  // This is not valid C so it can't be used in plain C...
  None = ZXing::Error::Type::None,
  Format = ZXing::Error::Type::Format,
  // ...
}
```

If there's project support, I would be interested in working on it (maybe under `wrappers/c/`?).

**Comment by axxel** [comment]
Thanks for bringing up this subject. I'd be happy to discuss this further, e.g. based on a concrete PR. `wrappers/c` sounds like a good place to start. Maybe in the end, we might even ship those c-ABI symbols directly with the main library. I don't yet see how a header-only version would work in your scenario. You still need a library file (like a `.so`) with c-mangled names to be accesible from your `go` program, right?

In terms of features, I would not require that wrapper to be feature complete from my side.

**Reply by siiky** [reply]
Thanks for your reply, it's encouraging! I wrote a sketch start of an implementation and opened https://github.com/zxing-cpp/zxing-cpp/pull/553 with it. My C++ noobness probably shows, I'll need some help there.

**Comment by sayantanrana** [comment]
Hello @siiky, could you please provide the url of the C-repo that you are developing?

**Reply by siiky** [reply]
Hey, you can find it on the opened WIP PR: https://github.com/zxing-cpp/zxing-cpp/pull/553

https://github.com/siiky/zxing-cpp/tree/c-wrapper

**Reply by sayantanrana** [reply]
Thanks

**Reply by axxel** [reply]
Note: #553 has been merged

**Comment by siiky** [comment]
May be closed now as #553 has been merged. Anyone willing to provide feedback is free to do it here, on the PR, or a new issue/discussion.
